<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ahmed M. Elhossiny">

<title>scRNAseq Data Processing and Integration â€“ Elhossiny et al, 2025 Code</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-c1fac2584b48ed01fb6e278e36375074.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Elhossiny et al, 2025 Code</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-scrnaseq-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">scRNAseq Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-scrnaseq-analysis">    
        <li>
    <a class="dropdown-item" href="./scRNAseq_data_acquisiton.html">
 <span class="dropdown-text">Data Acquisition</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_cellbender.html">
 <span class="dropdown-text">Ambient RNA Correction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_processing.html">
 <span class="dropdown-text">Data Processing and Integration</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_numbat.html">
 <span class="dropdown-text">CNV Inference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_fibro.html">
 <span class="dropdown-text">Firoblast Subpopulation Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_macs.html">
 <span class="dropdown-text">Macrophages Subpopulation Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNASeq_TCGA_Scoring.html">
 <span class="dropdown-text">TCGA-PAAD Scoring</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-visium-data-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Visium Data Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-visium-data-analysis">    
        <li>
    <a class="dropdown-item" href="./st_data_acquisiton.html">
 <span class="dropdown-text">Data Acquisition</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_processing.html">
 <span class="dropdown-text">Data Processing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_deconvolution.html">
 <span class="dropdown-text">Cell Type Deconvolution (RCTD)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_clustering.html">
 <span class="dropdown-text">Spatially-aware Clustering (BayesSpace)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_epi_analysis.qmd.qmd">
 <span class="dropdown-text">Epithelial Domains Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_nhood_analysis.html">
 <span class="dropdown-text">Neighborhood Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_stroma_analysis.html">
 <span class="dropdown-text">Stromal Domains Analysis</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-xenium-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Xenium Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-xenium-analysis">    
        <li>
    <a class="dropdown-item" href="./st_xenium_segmentation.html">
 <span class="dropdown-text">Resegmentation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_processing.html">
 <span class="dropdown-text">Data Processing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_regression.html">
 <span class="dropdown-text">Spatial Regression</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_viz.html">
 <span class="dropdown-text">Polygon Visualization</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-paper-figures" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Paper Figures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-paper-figures">    
        <li>
    <a class="dropdown-item" href="./fig2.html">
 <span class="dropdown-text">Fig 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig3.html">
 <span class="dropdown-text">Fig 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig4.html">
 <span class="dropdown-text">Fig 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig5.html">
 <span class="dropdown-text">Fig 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig6.html">
 <span class="dropdown-text">Fig 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig7.html">
 <span class="dropdown-text">Fig 7</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="https://pascadimaglianolab.shinyapps.io/PancAtlas/" target="_blank"> 
<span class="menu-text">Interactive Data Visualization</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup-and-environment-configuration" id="toc-setup-and-environment-configuration" class="nav-link active" data-scroll-target="#setup-and-environment-configuration">1 Setup and Environment Configuration</a></li>
  <li><a href="#data-loading-and-object-preparation" id="toc-data-loading-and-object-preparation" class="nav-link" data-scroll-target="#data-loading-and-object-preparation">2 Data Loading and Object Preparation</a></li>
  <li><a href="#quality-control-and-doublet-detection" id="toc-quality-control-and-doublet-detection" class="nav-link" data-scroll-target="#quality-control-and-doublet-detection">3 Quality Control and Doublet Detection</a>
  <ul class="collapse">
  <li><a href="#initial-qc-metrics-and-filtering" id="toc-initial-qc-metrics-and-filtering" class="nav-link" data-scroll-target="#initial-qc-metrics-and-filtering">3.1 Initial QC Metrics and Filtering</a></li>
  <li><a href="#doublet-detection-with-scrublet" id="toc-doublet-detection-with-scrublet" class="nav-link" data-scroll-target="#doublet-detection-with-scrublet">3.2 Doublet Detection with Scrublet</a></li>
  <li><a href="#apply-filters" id="toc-apply-filters" class="nav-link" data-scroll-target="#apply-filters">3.3 Apply Filters</a></li>
  </ul></li>
  <li><a href="#data-integration" id="toc-data-integration" class="nav-link" data-scroll-target="#data-integration">4 Data Integration</a>
  <ul class="collapse">
  <li><a href="#normalization-and-pre-processing" id="toc-normalization-and-pre-processing" class="nav-link" data-scroll-target="#normalization-and-pre-processing">4.1 Normalization and Pre-processing</a></li>
  <li><a href="#integration-with-scvi" id="toc-integration-with-scvi" class="nav-link" data-scroll-target="#integration-with-scvi">4.2 Integration with scVI</a></li>
  </ul></li>
  <li><a href="#dimensionality-reduction-and-clustering" id="toc-dimensionality-reduction-and-clustering" class="nav-link" data-scroll-target="#dimensionality-reduction-and-clustering">5 Dimensionality Reduction and Clustering</a>
  <ul class="collapse">
  <li><a href="#umap-embedding" id="toc-umap-embedding" class="nav-link" data-scroll-target="#umap-embedding">5.1 UMAP Embedding</a></li>
  <li><a href="#clustering-and-annotation" id="toc-clustering-and-annotation" class="nav-link" data-scroll-target="#clustering-and-annotation">5.2 Clustering and Annotation</a></li>
  </ul></li>
  <li><a href="#finalize-cell-type-annotations-and-generate-plots" id="toc-finalize-cell-type-annotations-and-generate-plots" class="nav-link" data-scroll-target="#finalize-cell-type-annotations-and-generate-plots">6 Finalize Cell Type Annotations and Generate Plots</a>
  <ul class="collapse">
  <li><a href="#renaming-clusters" id="toc-renaming-clusters" class="nav-link" data-scroll-target="#renaming-clusters">6.1 Renaming Clusters</a></li>
  </ul></li>
  <li><a href="#save-final-annotated-object" id="toc-save-final-annotated-object" class="nav-link" data-scroll-target="#save-final-annotated-object">7 Save Final Annotated Object</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">8 Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">scRNAseq Data Processing and Integration</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Ahmed M. Elhossiny <a href="mailto:hossiny@umich.edu" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="setup-and-environment-configuration" class="level2">
<h2 class="anchored" data-anchor-id="setup-and-environment-configuration">1 Setup and Environment Configuration</h2>
<p>Here, we load essential libraries and configure the <code>reticulate</code> package to use a specific Conda environment. This is crucial for running Python-based tools within the R environment. The exact conda environment used in this run can be installed using this <a href="envs/single_cell.yml">yml</a> file using <code>conda env create -f single_cell.yml</code></p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scCustomize)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(CellChat)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slurmR)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(numbat)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(unix)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">use_condaenv</span>(<span class="st">"path/to/conda/envs/single_cell/"</span>, <span class="at">required =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scanpy"</span>, <span class="at">convert =</span> F)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>anndata <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"anndata"</span>, <span class="at">convert =</span> F)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>scipy <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">'scipy'</span>, <span class="at">convert =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
<section id="data-loading-and-object-preparation" class="level2">
<h2 class="anchored" data-anchor-id="data-loading-and-object-preparation">2 Data Loading and Object Preparation</h2>
<p>We use Seurat V2 <span class="citation" data-cites="seurat">(<a href="#ref-seurat" role="doc-biblioref">Hao et al. 2023</a>)</span> We define the main output directory where all results and intermediate files will be saved. We then read the <code>sc_samples_manifest.xlsx</code> from the data directory, containing metadata for each sample, which will be used to annotate the cells.</p>
<p>We load CellBender-processed HDF5 count data for each sample using <code>Read_CellBender_h5_Mat</code> from <a href="https://github.com/samuel-marsh/scCustomize">scCustomize</a> R package, iterating through all sample IDs to read the data, create Seurat objects, and append metadata such as sample ID, patient ID, and tissue type from the manifest. Cell barcodes are prefixed with their sample IDs to ensure unique identifiers across samples, and all resulting Seurat objects are then merged into a single comprehensive object named ref.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>outputDir <span class="ot">&lt;-</span> <span class="st">'../outputs/scRNAseq_Analysis/'</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(outputDir, <span class="at">showWarnings =</span> F, <span class="at">recursive =</span> T)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>samples_info <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(<span class="st">"../data/sc_samples_manifest.xlsx"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">lapply</span>(samples_info<span class="sc">$</span>sample_id, <span class="cf">function</span>(x) {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  sample_info <span class="ot">&lt;-</span> samples_info <span class="sc">%&gt;%</span> <span class="fu">filter</span>(sample_id <span class="sc">==</span> x)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  seurat <span class="ot">&lt;-</span> <span class="fu">Read_CellBender_h5_Mat</span>(<span class="fu">paste0</span>(</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"../outputs/scRNASeq_Analysis/cellbender/"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    x,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"/"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    x,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"_filtered.h5"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">CreateSeuratObject</span>()</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  seurat<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> x</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  seurat<span class="sc">$</span>patient_id <span class="ot">&lt;-</span> sample_info<span class="sc">$</span>patient_id</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  seurat<span class="sc">$</span>tissue <span class="ot">&lt;-</span> sample_info<span class="sc">$</span>tissue</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  seurat <span class="ot">&lt;-</span> <span class="fu">RenameCells</span>(seurat, <span class="at">add.cell.id =</span> x)</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(seurat)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">reduce</span>(ref, merge)</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ref, <span class="at">assay =</span> <span class="st">'RNA'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
<section id="quality-control-and-doublet-detection" class="level2">
<h2 class="anchored" data-anchor-id="quality-control-and-doublet-detection">3 Quality Control and Doublet Detection</h2>
<section id="initial-qc-metrics-and-filtering" class="level3">
<h3 class="anchored" data-anchor-id="initial-qc-metrics-and-filtering">3.1 Initial QC Metrics and Filtering</h3>
<p>We calculate the percentage of mitochondrial genes for each cell (<code>percent.mt</code>), a key quality control metric for cell quality. We also perform an initial filtering step to remove cells and genes that have zero counts across the entire dataset.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ref[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(<span class="at">object =</span> ref, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> ref[,<span class="sc">!</span>(<span class="fu">colSums</span>(<span class="fu">GetAssayData</span>(ref, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)) <span class="sc">==</span> <span class="dv">0</span>)]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> ref[<span class="sc">!</span>(<span class="fu">rowSums</span>(<span class="fu">GetAssayData</span>(ref, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)) <span class="sc">==</span> <span class="dv">0</span>),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="doublet-detection-with-scrublet" class="level3">
<h3 class="anchored" data-anchor-id="doublet-detection-with-scrublet">3.2 Doublet Detection with Scrublet</h3>
<p>We run Scrublet <span class="citation" data-cites="scrublet">(<a href="#ref-scrublet" role="doc-biblioref">Wolock, Lopez, and Klein 2019</a>)</span>, a Python-based tool, to identify potential doublets (two or more cells captured as a single droplet). We run it on each sample separately, then the results are transferred back to R and added as metadata to the main Seurat object.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>scrublet_res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(ref<span class="sc">$</span>sample_id), <span class="cf">function</span>(x) {</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(x)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  sample <span class="ot">&lt;-</span> <span class="fu">subset</span>(ref, <span class="at">subset =</span> sample_id <span class="sc">==</span> x)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  adata <span class="ot">&lt;-</span> anndata<span class="sc">$</span><span class="fu">AnnData</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> scipy<span class="sc">$</span>sparse<span class="sc">$</span><span class="fu">csr_matrix</span>(</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>      Matrix<span class="sc">::</span><span class="fu">t</span>(<span class="fu">LayerData</span>(sample, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">"counts"</span>))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  sc<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">scrublet</span>(adata)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  scrublet_res <span class="ot">&lt;-</span> <span class="fu">py_to_r</span>(adata<span class="sc">$</span>obs) <span class="sc">%&gt;%</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(doublet_score, predicted_doublet)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(scrublet_res) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sample)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  scrublet_res<span class="sc">$</span>predicted_doublet <span class="ot">&lt;-</span> <span class="fu">unlist</span>(scrublet_res<span class="sc">$</span>predicted_doublet)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(scrublet_res)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>()</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(ref, scrublet_res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="apply-filters" class="level3">
<h3 class="anchored" data-anchor-id="apply-filters">3.3 Apply Filters</h3>
<p>We apply standard quality control filters, removing cells with high mitochondrial content (<code>percent.mt &lt; 15</code>) or a low number of detected features (<code>nFeature_RNA &gt; 200</code>).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">subset</span>(ref, <span class="at">subset =</span> percent.mt <span class="sc">&lt;</span> <span class="dv">15</span> <span class="sc">&amp;</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
</section>
<section id="data-integration" class="level2">
<h2 class="anchored" data-anchor-id="data-integration">4 Data Integration</h2>
<section id="normalization-and-pre-processing" class="level3">
<h3 class="anchored" data-anchor-id="normalization-and-pre-processing">4.1 Normalization and Pre-processing</h3>
<p>Before integration, we perform standard Seurat pre-processing steps. The RNA assay is split by sample to prepare for integration, the data is log-normalized, highly variable features are identified, the data is scaled, and Principal Component Analysis (PCA) is run.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ref[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ref[[<span class="st">"RNA"</span>]], <span class="at">f =</span> ref<span class="sc">$</span>sample_id)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(ref)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(ref)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(ref)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ref)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="integration-with-scvi" class="level3">
<h3 class="anchored" data-anchor-id="integration-with-scvi">4.2 Integration with scVI</h3>
<p>To correct for batch effects between different samples, we perform data integration using scVI <span class="citation" data-cites="scvi">(<a href="#ref-scvi" role="doc-biblioref">Lopez et al. 2018</a>)</span>, A deep learning-based method implemented via <a href="https://github.com/satijalab/seurat-wrappers"><code>SeuratWrappers</code></a>. A new dimensional reduction slots (<code>integrated.scvi</code>) is created in the Seurat object.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> ref,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> scVIIntegration,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">conda_env =</span> <span class="st">"path/to/conda/envs/single_cell/"</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">new.reduction =</span> <span class="st">"integrated.scvi"</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
</section>
<section id="dimensionality-reduction-and-clustering" class="level2">
<h2 class="anchored" data-anchor-id="dimensionality-reduction-and-clustering">5 Dimensionality Reduction and Clustering</h2>
<section id="umap-embedding" class="level3">
<h3 class="anchored" data-anchor-id="umap-embedding">5.1 UMAP Embedding</h3>
<p>We then calculate UMAP (Uniform Manifold Approximation and Projection) coordinates based on the <code>scVI</code> reduction.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ref)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  ref,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">'integrated.scvi'</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="clustering-and-annotation" class="level3">
<h3 class="anchored" data-anchor-id="clustering-and-annotation">5.2 Clustering and Annotation</h3>
<p>Using the scVI-based reduction, we generate graph and perform graph-based clustering. A range of resolutions is tested (from 0.05 to 1.0) to explore different clustering granularities. The clustree <span class="citation" data-cites="clustree">(<a href="#ref-clustree" role="doc-biblioref">Zappia and Oshlack 2018</a>)</span> R package helps visualize how cells move between clusters at different resolutions, aiding in the selection of an optimal resolution.</p>
<p>A final clustering resolution of <code>0.2</code> is chosen. We then find markers for these clusters and load a pre-defined annotation table from an Excel file <code>scvi_cluster_annotation.xlsx</code> in <code>outputs/scRNAseq_Analysis/</code> directory, created to assign cell type labels to the clusters.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ref, <span class="at">reduction =</span> <span class="st">'integrated.scvi'</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="fl">0.05</span>, <span class="at">to =</span> <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.05</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (x <span class="cf">in</span> res) {</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  ref <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ref, <span class="at">resolution =</span> x, <span class="at">cluster.name =</span> <span class="fu">paste0</span>(<span class="st">"res."</span>, x))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">clustree</span>(ref, <span class="at">prefix =</span> <span class="st">'res.'</span>, <span class="at">layout =</span> <span class="st">"sugiyama"</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ref, <span class="at">cluster.name =</span> <span class="st">'scvi_clusters'</span>, <span class="at">resolution =</span> <span class="fl">0.2</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>ref<span class="sc">@</span>misc<span class="sc">$</span>markers_scvi <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(ref, <span class="at">assay =</span> <span class="st">'RNA'</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>ref<span class="sc">@</span>misc<span class="sc">$</span>markers_scvi<span class="sc">$</span>score <span class="ot">&lt;-</span> ref<span class="sc">@</span>misc<span class="sc">$</span>markers_scvi<span class="sc">$</span>avg_log2FC <span class="sc">*</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  ref<span class="sc">@</span>misc<span class="sc">$</span>markers_scvi<span class="sc">$</span>pct<span class="fl">.1</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>annotation <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"outputs/scRNAseq_Analysis/scvi_cluster_annotation.xlsx"</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>CellTypes <span class="ot">&lt;-</span> <span class="fu">setNames</span>(annotation<span class="sc">$</span>celltype, annotation<span class="sc">$</span>cluster)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ref) <span class="ot">&lt;-</span> ref<span class="sc">$</span>scvi_clusters</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(ref, CellTypes)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>ref<span class="sc">$</span>CellTypes <span class="ot">&lt;-</span> <span class="fu">Idents</span>(ref)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="finalize-cell-type-annotations-and-generate-plots" class="level2">
<h2 class="anchored" data-anchor-id="finalize-cell-type-annotations-and-generate-plots">6 Finalize Cell Type Annotations and Generate Plots</h2>
<section id="renaming-clusters" class="level3">
<h3 class="anchored" data-anchor-id="renaming-clusters">6.1 Renaming Clusters</h3>
<p>The <code>CellTypes</code> metadata column is converted to a factor with a specific level order. This ensures that cell types appear in a consistent and logical order in all subsequent plots. A list of canonical marker genes is defined for visualization.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ref<span class="sc">$</span>main_annotation_scvi <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  ref<span class="sc">$</span>main_annotation_scvi,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CHRM3+"</span>, <span class="st">"Acinar"</span>, <span class="st">"Acinar/ADM"</span>, <span class="st">"ADM"</span>, <span class="st">"Ducts"</span>, <span class="st">"Tumor_Epithelial"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Macrophages"</span>, <span class="st">"Granulocytes"</span>, <span class="st">"T_Cells"</span>, <span class="st">"NK_Cells"</span>, <span class="st">"B_Cells"</span>, <span class="st">"Plasma"</span>, <span class="st">"Mast_Cells"</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Endothelial"</span>, <span class="st">"Lymphatic_Endothelial"</span>, <span class="st">"Endocrine"</span>, <span class="st">"Fibroblasts"</span>, <span class="st">"Pericytes"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Neurons"</span>, <span class="st">"Muscle"</span>))</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"CHRM3"</span>,<span class="st">"MECOM"</span>,<span class="st">"PRSS1"</span>,<span class="st">"SPINK1"</span>,<span class="st">"AMY2A"</span>,<span class="st">"SOX9"</span>,<span class="st">"MUC6"</span>,<span class="st">"FXYD2"</span>,<span class="st">"CFTR"</span>,<span class="st">"AQP1"</span>,<span class="st">"BICC1"</span>,<span class="st">"MUC5B"</span>,<span class="st">"CRISP3"</span>,<span class="st">"MSLN"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"CEACAM6"</span>,<span class="st">"KRT19"</span>,<span class="st">"KRT17"</span>,<span class="st">"LAMB3"</span>,<span class="st">"GPRC5A"</span>,<span class="st">"KLK10"</span>,<span class="st">"APOE"</span>,<span class="st">"CD68"</span>,<span class="st">"S100A8"</span>,<span class="st">"S100A9"</span>,<span class="st">"FCGR3B"</span>,<span class="st">"CSF3R"</span>,<span class="st">"CD3E"</span>,<span class="st">"IL7R"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"NKG7"</span>,<span class="st">"GNLY"</span>,<span class="st">"MS4A1"</span>,<span class="st">"CD79A"</span>,<span class="st">"JCHAIN"</span>,<span class="st">"IGKC"</span>,<span class="st">"CPA3"</span>,<span class="st">"TPSAB1"</span>,<span class="st">"VWF"</span>,<span class="st">"CDH5"</span>,<span class="st">"NTS"</span>,<span class="st">"MMRN1"</span>,<span class="st">"PKHD1L1"</span>,<span class="st">"INS"</span>,<span class="st">"GCG"</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"CHGA"</span>,<span class="st">"PDGFRA"</span>,<span class="st">"COL1A1"</span>,<span class="st">"LUM"</span>,<span class="st">"DCN"</span>,<span class="st">"RGS5"</span>,<span class="st">"TAGLN"</span>,<span class="st">"PLP1"</span>,<span class="st">"NRXN1"</span>,<span class="st">"NCAM2"</span>,<span class="st">"TRDN"</span>,<span class="st">"MYBPC1"</span>,<span class="st">"NEB"</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"objects/scRef.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot_scCustom</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  ref,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'CellTypes'</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">'umap.scvi'</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot_scCustom</span>(</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  ref,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> features,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'CellTypes'</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">flip_axes =</span> T,</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">x_lab_rotate =</span> T</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">'bold'</span>),</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">'bold'</span>)</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="save-final-annotated-object" class="level2">
<h2 class="anchored" data-anchor-id="save-final-annotated-object">7 Save Final Annotated Object</h2>
<p>The fully processed, integrated, and annotated Seurat object is saved to a file using <code>qsave</code>. This final object can be easily loaded for any further downstream analyses.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(ref, <span class="st">"outputs/scRNAseq_Analysis/scRef.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The final processed file is deposited in Zenodo and can be downloaded from <a href="">here</a></p>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">8 Session Info</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-seurat" class="csl-entry" role="listitem">
Hao, Yuhan, Tim Stuart, Madeline H Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, et al. 2023. <span>â€œDictionary Learning for Integrative, Multimodal and Scalable Single-Cell Analysis.â€</span> <em>Nature Biotechnology</em>. <a href="https://doi.org/10.1038/s41587-023-01767-y">https://doi.org/10.1038/s41587-023-01767-y</a>.
</div>
<div id="ref-scvi" class="csl-entry" role="listitem">
Lopez, Romain, Jeffrey Regier, Michael B Cole, Michael I Jordan, and Nir Yosef. 2018. <span>â€œDeep Generative Modeling for Single-Cell Transcriptomics.â€</span> <em>Nature Methods</em> 15 (12): 1053â€“58.
</div>
<div id="ref-scrublet" class="csl-entry" role="listitem">
Wolock, Samuel L., Romain Lopez, and Allon M. Klein. 2019. <span>â€œScrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data.â€</span> <em>Cell Systems</em> 8 (4): 281â€“291.e9. https://doi.org/<a href="https://doi.org/10.1016/j.cels.2018.11.005">https://doi.org/10.1016/j.cels.2018.11.005</a>.
</div>
<div id="ref-clustree" class="csl-entry" role="listitem">
Zappia, Luke, and Alicia Oshlack. 2018. <span>â€œClustering Trees: A Visualization for Evaluating Clusterings at Multiple Resolutions.â€</span> <em>Gigascience</em> 7 (7): giy083.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>