<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ahmed M. Elhossiny">

<title>Spatial Transcriptomics Stromal Domains Analysis â€“ PanIN &amp; PDAC Spatial Transcriptomics Paper Code</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-c1fac2584b48ed01fb6e278e36375074.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">PanIN &amp; PDAC Spatial Transcriptomics Paper Code</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-scrnaseq-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">scRNAseq Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-scrnaseq-analysis">    
        <li>
    <a class="dropdown-item" href="./scRNAseq_data_acquisiton.html">
 <span class="dropdown-text">Data Acquisition</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_cellbender.html">
 <span class="dropdown-text">Ambient RNA Correction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_processing.html">
 <span class="dropdown-text">Data Processing and Integration</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_numbat.html">
 <span class="dropdown-text">CNV Inference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_fibro.html">
 <span class="dropdown-text">Firoblast Subpopulation Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNAseq_macs.html">
 <span class="dropdown-text">Macrophages Subpopulation Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./scRNASeq_TCGA_Scoring.html">
 <span class="dropdown-text">TCGA-PAAD Scoring</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-visium-data-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Visium Data Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-visium-data-analysis">    
        <li>
    <a class="dropdown-item" href="./st_data_acquisiton.html">
 <span class="dropdown-text">Data Acquisition</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_processing.html">
 <span class="dropdown-text">Data Processing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_deconvolution.html">
 <span class="dropdown-text">Cell Type Deconvolution (RCTD)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_clustering.html">
 <span class="dropdown-text">Spatially-aware Clustering (BayesSpace)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_epi_analysis.qmd.qmd">
 <span class="dropdown-text">Epithelial Domains Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_nhood_analysis.html">
 <span class="dropdown-text">Neighborhood Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_stroma_analysis.html">
 <span class="dropdown-text">Stromal Domains Analysis</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-xenium-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Xenium Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-xenium-analysis">    
        <li>
    <a class="dropdown-item" href="./st_xenium_segmentation.html">
 <span class="dropdown-text">Resegmentation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_processing.html">
 <span class="dropdown-text">Data Processing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_regression.html">
 <span class="dropdown-text">Spatial Regression</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./st_xenium_viz.html">
 <span class="dropdown-text">Polygon Visualization</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-paper-figures" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Paper Figures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-paper-figures">    
        <li>
    <a class="dropdown-item" href="./fig2.html">
 <span class="dropdown-text">Fig 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig3.html">
 <span class="dropdown-text">Fig 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig4.html">
 <span class="dropdown-text">Fig 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig5.html">
 <span class="dropdown-text">Fig 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig6.html">
 <span class="dropdown-text">Fig 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./fig7.html">
 <span class="dropdown-text">Fig 7</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="https://pascadimaglianolab.shinyapps.io/PancAtlas/" target="_blank"> 
<span class="menu-text">Interactive Data Visualization</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup-and-import-data" id="toc-setup-and-import-data" class="nav-link active" data-scroll-target="#setup-and-import-data">1 Setup and import data</a></li>
  <li><a href="#subsetting-fibroblasts-enriched-domains" id="toc-subsetting-fibroblasts-enriched-domains" class="nav-link" data-scroll-target="#subsetting-fibroblasts-enriched-domains">2 Subsetting Fibroblasts-enriched domains</a></li>
  <li><a href="#generating-scref-bayesprism-reference" id="toc-generating-scref-bayesprism-reference" class="nav-link" data-scroll-target="#generating-scref-bayesprism-reference">3 Generating scRef BayesPrism Reference</a></li>
  <li><a href="#extracting-fibroblast-specific-expression-from-spatial-domains" id="toc-extracting-fibroblast-specific-expression-from-spatial-domains" class="nav-link" data-scroll-target="#extracting-fibroblast-specific-expression-from-spatial-domains">4 Extracting Fibroblast-specific expression from spatial domains</a></li>
  <li><a href="#pseudobulk-dge-analysis" id="toc-pseudobulk-dge-analysis" class="nav-link" data-scroll-target="#pseudobulk-dge-analysis">5 Pseudobulk DGE analysis</a></li>
  <li><a href="#hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes" id="toc-hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes" class="nav-link" data-scroll-target="#hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes">6 Hierarchical Deconvolution using scRNAseq-defined fibroblast subtypes</a>
  <ul class="collapse">
  <li><a href="#generating-fibroblasts-scrnaseq-bayesprism-reference" id="toc-generating-fibroblasts-scrnaseq-bayesprism-reference" class="nav-link" data-scroll-target="#generating-fibroblasts-scrnaseq-bayesprism-reference">6.1 Generating fibroblasts scRNAseq BayesPrism Reference</a></li>
  <li><a href="#running-bayesprism" id="toc-running-bayesprism" class="nav-link" data-scroll-target="#running-bayesprism">6.2 Running BayesPrism</a></li>
  <li><a href="#visualizing-fibroblast-subtype-fractions" id="toc-visualizing-fibroblast-subtype-fractions" class="nav-link" data-scroll-target="#visualizing-fibroblast-subtype-fractions">6.3 Visualizing Fibroblast Subtype Fractions</a></li>
  </ul></li>
  <li><a href="#hierarchical-deconvolution-using-scrnaseq-defined-macrophage-subtypes" id="toc-hierarchical-deconvolution-using-scrnaseq-defined-macrophage-subtypes" class="nav-link" data-scroll-target="#hierarchical-deconvolution-using-scrnaseq-defined-macrophage-subtypes">7 Hierarchical Deconvolution using scRNAseq-defined macrophage subtypes</a>
  <ul class="collapse">
  <li><a href="#extracting-macrophage-specific-expression-from-spatial-domains" id="toc-extracting-macrophage-specific-expression-from-spatial-domains" class="nav-link" data-scroll-target="#extracting-macrophage-specific-expression-from-spatial-domains">7.1 Extracting Macrophage-specific expression from spatial domains</a></li>
  <li><a href="#hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes-1" id="toc-hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes-1" class="nav-link" data-scroll-target="#hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes-1">7.2 Hierarchical Deconvolution using scRNAseq-defined fibroblast subtypes</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Spatial Transcriptomics Stromal Domains Analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Ahmed M. Elhossiny <a href="mailto:hossiny@umich.edu" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<p>Here we will be analyzing the stroma-enriched regions in the spatial transcriptomics data. We will be using the same hierarchical deconvolution used for the Stromal domains analysis using BayesPrism <span class="citation" data-cites="bayesprism">(<a href="#ref-bayesprism" role="doc-biblioref">Chu et al. 2022</a>)</span>.</p>
<section id="setup-and-import-data" class="level2">
<h2 class="anchored" data-anchor-id="setup-and-import-data">1 Setup and import data</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BayesPrism)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(qs)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(spatstat)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(DESeq2)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(PCAtools)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidytext)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(GSVA)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>samples_info <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(<span class="st">"../data/visium_samples_manifest.xlsx"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing data ----------------------------------------------------------</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>bp_scRef <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"outputs/Deconvolution/BayesPrism_scRef.qs"</span>)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>fibro <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"outputs/scRNAseq_Analysis/fibro_all.qs"</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>macs <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"outputs/scRNAseq_Analysis/macs_all.qs"</span>)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>merged <span class="ot">&lt;-</span> qs<span class="sc">::</span><span class="fu">qread</span>(<span class="st">"outputs/Spatial_Clustering/tumor_tumoradj_GoL_samples_annotated_v2_lite_new_deconv.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="subsetting-fibroblasts-enriched-domains" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-fibroblasts-enriched-domains">2 Subsetting Fibroblasts-enriched domains</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">scPalette</span>(<span class="fu">length</span>(<span class="fu">levels</span>(merged<span class="sc">$</span>main_annotation)))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cols) <span class="ot">&lt;-</span> <span class="fu">levels</span>(merged<span class="sc">$</span>main_annotation)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>all_fibro <span class="ot">&lt;-</span> <span class="fu">subset</span>(merged,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">subset =</span> main_annotation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Fibro1"</span>, <span class="st">"Fibro2"</span>, <span class="st">"Fibro3"</span>, <span class="st">"Fibro4(Hi_Mt)"</span>))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>histogram_samples <span class="ot">&lt;-</span> <span class="fu">ClusterDistrBar</span>(all_fibro<span class="sc">$</span>tissue, all_fibro<span class="sc">$</span>main_annotation,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">cols =</span> cols, <span class="at">flip =</span> F,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">border =</span> <span class="st">'white'</span>, <span class="at">reverse_order =</span> F) <span class="sc">+</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="cn">NULL</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"% of Total Fibroblast-Enriched Domains"</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>stroma <span class="ot">&lt;-</span> <span class="fu">subset</span>(merged, <span class="at">subset =</span> main_annotation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Fibro2"</span>, <span class="st">"Fibro3"</span>))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>stroma <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(stroma, <span class="at">reduction =</span> <span class="st">'integrated.rpca'</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(stroma, <span class="st">"outputs/Fibroblasts_Analysis/spatial_stroma.qs"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>samples_keep <span class="ot">&lt;-</span> <span class="fu">table</span>(stroma<span class="sc">$</span>sample_id, stroma<span class="sc">$</span>main_annotation) <span class="sc">%&gt;%</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Freq <span class="sc">&gt;</span> <span class="dv">50</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>stroma<span class="sc">$</span>sample_main_annotation <span class="ot">&lt;-</span> <span class="fu">paste0</span>(stroma<span class="sc">$</span>sample_id, <span class="st">"_"</span>, stroma<span class="sc">$</span>main_annotation)</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>stroma <span class="ot">&lt;-</span> <span class="fu">subset</span>(stroma, <span class="at">subset =</span> sample_main_annotation <span class="sc">%in%</span> <span class="fu">paste0</span>(samples_keep<span class="sc">$</span>Var1,<span class="st">"_"</span>, samples_keep<span class="sc">$</span>Var2))</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>stroma_aggregate <span class="ot">&lt;-</span> <span class="fu">AggregateExpression</span>(stroma, <span class="at">assays =</span> <span class="st">'Spatial'</span>, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"main_annotation"</span>, <span class="st">"tissue"</span>, <span class="st">"sample_id"</span>), <span class="at">return.seurat =</span> T)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>stroma_aggregate<span class="sc">$</span>main_annotation_tissue <span class="ot">&lt;-</span> <span class="fu">paste0</span>(stroma_aggregate<span class="sc">$</span>main_annotation, <span class="st">"_"</span>, stroma_aggregate<span class="sc">$</span>tissue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="generating-scref-bayesprism-reference" class="level2">
<h2 class="anchored" data-anchor-id="generating-scref-bayesprism-reference">3 Generating scRef BayesPrism Reference</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating scRef ----------------------------------------------------------</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>scRef <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"outputs/scRNAseq_Analysis/scRef_filtered_integrated_annotated.qs"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>scRef <span class="ot">&lt;-</span> <span class="fu">subset</span>(scRef, <span class="at">subset =</span> tissue <span class="sc">!=</span> <span class="st">'AdjNormal'</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>scRef <span class="ot">&lt;-</span> <span class="fu">subset</span>(scRef, <span class="at">subset =</span> main_annotation_scvi <span class="sc">!=</span> <span class="st">'T_Cells(Hi_Rb)'</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>scRef<span class="sc">$</span>main_annotation_scvi_tissue <span class="ot">&lt;-</span> <span class="fu">paste0</span>(scRef<span class="sc">$</span>main_annotation_scvi, <span class="st">"_"</span>, scRef<span class="sc">$</span>tissue)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> scRef<span class="sc">@</span>meta.data</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>sc.dat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">GetAssayData</span>(scRef, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered <span class="ot">&lt;-</span> <span class="fu">cleanup.genes</span>(<span class="at">input=</span><span class="fu">as.matrix</span>(sc.dat),</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>                                <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>                                <span class="at">species=</span><span class="st">"hs"</span>,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                                <span class="at">gene.group=</span><span class="fu">c</span>(<span class="st">"Rb"</span>,<span class="st">"Mrp"</span>,<span class="st">"other_Rb"</span>,<span class="st">"chrM"</span>,<span class="st">"MALAT1"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>) ,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>                                <span class="at">exp.cells=</span><span class="dv">10</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered.pc <span class="ot">&lt;-</span>  <span class="fu">select.gene.type</span>(sc.dat.filtered, <span class="at">gene.type =</span> <span class="st">"protein_coding"</span>)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(sc.dat.filtered); <span class="fu">gc</span>()</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>scRef <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">sc.dat =</span> sc.dat.filtered.pc, <span class="at">metadata =</span> metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="extracting-fibroblast-specific-expression-from-spatial-domains" class="level2">
<h2 class="anchored" data-anchor-id="extracting-fibroblast-specific-expression-from-spatial-domains">4 Extracting Fibroblast-specific expression from spatial domains</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ct_mtx <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">t</span>(<span class="fu">GetAssayData</span>(stroma_aggregate, <span class="at">assay =</span> <span class="st">'Spatial'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>intersect_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(ct_mtx), <span class="fu">colnames</span>(bp_scRef<span class="sc">$</span>sc.dat))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>myPrism <span class="ot">&lt;-</span> <span class="fu">new.prism</span>(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference=</span>bp_scRef<span class="sc">$</span>sc.dat[,intersect_genes],</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">mixture=</span>ct_mtx[,intersect_genes],</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.type.labels =</span> bp_scRef<span class="sc">$</span>metadata<span class="sc">$</span>main_annotation_scvi,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.state.labels =</span> bp_scRef<span class="sc">$</span>metadata<span class="sc">$</span>main_annotation_scvi_tissue,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">key =</span> <span class="st">'Tumor_Epithelial'</span>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>bp_res <span class="ot">&lt;-</span> <span class="fu">run.prism</span>(myPrism, <span class="at">n.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>stroma_aggregate[[<span class="st">'fibroexp'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.exp</span>(bp_res, <span class="at">state.or.type =</span> <span class="st">'type'</span>, <span class="at">cell.name =</span> <span class="st">'Fibroblasts'</span>))))</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>stroma_aggregate[[<span class="st">'celltype_fractions'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.fraction</span>(bp_res, <span class="at">which.theta =</span> <span class="st">'final'</span>, <span class="at">state.or.type =</span> <span class="st">'type'</span>)), <span class="at">digits =</span> <span class="dv">2</span>))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>stroma_aggregate<span class="sc">@</span>assays<span class="sc">$</span>fibroexp<span class="sc">$</span>data <span class="ot">&lt;-</span> <span class="fu">vst</span>(<span class="fu">as.matrix</span>(<span class="fu">GetAssayData</span>(stroma_aggregate, <span class="at">assay =</span> <span class="st">'fibroexp'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="pseudobulk-dge-analysis" class="level2">
<h2 class="anchored" data-anchor-id="pseudobulk-dge-analysis">5 Pseudobulk DGE analysis</h2>
<p>We will use DESeq2 <span class="citation" data-cites="DESeq2">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> for differential expression analysis of the fibroblast-specific expression from Fibro2 and Fibro3 spatial domains.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>ct_data <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(stroma_aggregate, <span class="at">assay =</span> <span class="st">'fibroexp'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">&lt;-</span> stroma_aggregate<span class="sc">@</span>meta.data</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>run_id <span class="ot">&lt;-</span> samples_info <span class="sc">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(patient_id, run_id) <span class="sc">%&gt;%</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">sample_id =</span> patient_id) <span class="sc">%&gt;%</span> </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sample_id =</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">_"</span>, <span class="st">"-"</span>, sample_id)) <span class="sc">%&gt;%</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">&lt;-</span> <span class="fu">merge</span>(colData, run_id, <span class="at">by =</span> <span class="st">'sample_id'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">'orig.ident'</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>vst <span class="ot">&lt;-</span> <span class="fu">vst</span>(<span class="fu">as.matrix</span>(ct_data))</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">pca</span>(vst, colData[<span class="fu">colnames</span>(vst),])</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(pca, <span class="st">"outputs/Fibroblasts_Analysis/fibro_aggregate_pca.qs"</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pca,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">colby =</span> <span class="st">'main_annotation'</span>,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">lab =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">shape =</span> <span class="st">'tissue'</span>,</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">encircle =</span> F, <span class="at">legendTitleSize =</span> <span class="dv">12</span>,</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>       <span class="at">legendPosition =</span> <span class="st">'right'</span>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>colData<span class="sc">$</span>main_annotation <span class="ot">&lt;-</span> <span class="fu">factor</span>(colData<span class="sc">$</span>main_annotation, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Fibro3"</span>, <span class="st">"Fibro2"</span>))</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> ct_data,</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> colData[<span class="fu">colnames</span>(ct_data),],</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design =</span> <span class="sc">~</span> sample_id <span class="sc">+</span> main_annotation)</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>stroma2_v_stroma3 <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef =</span> <span class="st">'main_annotation_Fibro2_vs_Fibro3'</span>, <span class="at">type =</span> <span class="st">'ashr'</span>) <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>()</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  stroma3_v_stroma2,</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(stroma3_v_stroma2),</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  <span class="st">'log2FoldChange'</span>,</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  <span class="st">'padj'</span>,</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">pCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">'Fibro3 vs Fibro2'</span>,</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">subtitle =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">drawConnectors =</span> T,</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">labSize =</span> <span class="dv">2</span>,</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">titleLabSize =</span> <span class="dv">15</span>,</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">legendPosition =</span> <span class="st">'none'</span>,</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">caption =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>  <span class="co"># border = 'full',</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>  <span class="at">axisLabSize =</span> <span class="dv">10</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes" class="level2">
<h2 class="anchored" data-anchor-id="hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes">6 Hierarchical Deconvolution using scRNAseq-defined fibroblast subtypes</h2>
<p>Here we will be using Fibroblast subtypes defined from scRNAseq data to dissect the composition of Fibro2 and Fibro3 spatial domains.</p>
<section id="generating-fibroblasts-scrnaseq-bayesprism-reference" class="level3">
<h3 class="anchored" data-anchor-id="generating-fibroblasts-scrnaseq-bayesprism-reference">6.1 Generating fibroblasts scRNAseq BayesPrism Reference</h3>
<p>We will remove the AdjNormal tissue samples from scRNAseq reference in the analysis as we described in <a href="./st_epi_analysis.html">Epithelial Domains Analysis</a></p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>fibro <span class="ot">&lt;-</span> <span class="fu">subset</span>(fibro, <span class="at">subset =</span> tissue <span class="sc">!=</span> <span class="st">'AdjNormal'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> fibro<span class="sc">@</span>meta.data</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>fibro_subtypes_tissue <span class="ot">&lt;-</span> <span class="fu">paste0</span>(metadata<span class="sc">$</span>fibro_subtypes, <span class="st">"_"</span>, metadata<span class="sc">$</span>tissue)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>sc.dat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">GetAssayData</span>(fibro, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered <span class="ot">&lt;-</span> <span class="fu">cleanup.genes</span>(<span class="at">input=</span><span class="fu">as.matrix</span>(sc.dat),</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">species=</span><span class="st">"hs"</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">gene.group=</span><span class="fu">c</span>( <span class="st">"Rb"</span>,<span class="st">"Mrp"</span>,<span class="st">"other_Rb"</span>,<span class="st">"chrM"</span>,<span class="st">"MALAT1"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>) ,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">exp.cells=</span><span class="dv">5</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered.pc <span class="ot">&lt;-</span>  <span class="fu">select.gene.type</span>(sc.dat.filtered, <span class="at">gene.type =</span> <span class="st">"protein_coding"</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>fibro_scRef <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">sc.dat =</span> sc.dat.filtered.pc, <span class="at">metadata =</span> metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="running-bayesprism" class="level3">
<h3 class="anchored" data-anchor-id="running-bayesprism">6.2 Running BayesPrism</h3>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>ct_mtx <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">t</span>(<span class="fu">GetAssayData</span>(stroma_aggregate, <span class="at">assay =</span> <span class="st">'fibroexp'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>intersect_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(ct_mtx), <span class="fu">colnames</span>(fibro_scRef<span class="sc">$</span>sc.dat))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>myPrism <span class="ot">&lt;-</span> <span class="fu">new.prism</span>(</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference=</span>fibro_scRef<span class="sc">$</span>sc.dat[,intersect_genes],</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">mixture=</span>ct_mtx[,intersect_genes],</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.type.labels =</span> fibro_scRef<span class="sc">$</span>metadata<span class="sc">$</span>fibro_subtypes,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.state.labels =</span> fibro_scRef<span class="sc">$</span>metadata<span class="sc">$</span>fibro_subtypes_tissue,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">key =</span> <span class="cn">NULL</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>fibro_bp_res <span class="ot">&lt;-</span> <span class="fu">run.prism</span>(myPrism, <span class="at">n.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(fibro_bp_res, <span class="st">"outputs/Fibroblasts_Analysis/fibro_subtypes_F2F3.qs"</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>stroma_aggregate[[<span class="st">'subtype_fractions'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.fraction</span>(fibro_bp_res, <span class="at">which.theta =</span> <span class="st">'final'</span>, <span class="at">state.or.type =</span> <span class="st">'type'</span>)), <span class="at">digits =</span> <span class="dv">2</span>))</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(stroma_aggregate) <span class="ot">&lt;-</span>  <span class="st">'subtype_fractions'</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(stroma_aggregate, <span class="st">"outputs/Fibroblasts_Analysis/fibroblast_subtypes_deconvolution_F2F3.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="visualizing-fibroblast-subtype-fractions" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-fibroblast-subtype-fractions">6.3 Visualizing Fibroblast Subtype Fractions</h3>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>avg <span class="ot">&lt;-</span> <span class="fu">AverageExpression</span>(stroma_aggregate, <span class="at">assays =</span> <span class="st">'subtype_fractions'</span>, <span class="at">group.by =</span> <span class="st">'main_annotation_tissue'</span>)<span class="sc">$</span>subtype_fractions</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>avg <span class="ot">&lt;-</span> avg[<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">7</span>,<span class="dv">1</span>,<span class="dv">3</span>),<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>, <span class="dv">3</span>,<span class="dv">1</span>)]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">as.data.frame</span>(avg), <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F, <span class="at">color =</span> <span class="fu">viridis</span>(<span class="at">n =</span> <span class="dv">100</span>, <span class="at">option =</span> <span class="st">'B'</span>, <span class="at">direction =</span> <span class="dv">1</span>), <span class="at">cellwidth =</span> <span class="dv">40</span>, <span class="at">cellheight =</span> <span class="dv">20</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>ann <span class="ot">&lt;-</span> stroma_aggregate<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(new_annotation)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>ann<span class="sc">$</span>new_annotation <span class="ot">&lt;-</span> <span class="fu">factor</span>(ann<span class="sc">$</span>new_annotation,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Fibro3_GoL"</span>, <span class="st">"Fibro3_Tumor"</span>, <span class="st">"Fibro2_Tumor"</span>))</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>fractions <span class="ot">&lt;-</span> <span class="fu">get.fraction</span>(fibro_bp_res, <span class="at">which.theta =</span> <span class="st">'final'</span>, <span class="at">state.or.type =</span> <span class="st">'type'</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">t</span>(fractions)[<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">7</span>,<span class="dv">1</span>,<span class="dv">3</span>),<span class="fu">order</span>(ann<span class="sc">$</span>new_annotation)], </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">cellwidth =</span> <span class="dv">10</span>, <span class="at">cellheight =</span> <span class="dv">10</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F, <span class="at">show_colnames =</span> F, <span class="at">annotation_col =</span> ann, <span class="fu">viridis</span>(<span class="at">option =</span> <span class="st">'B'</span>, <span class="at">n =</span> <span class="dv">100</span>, <span class="at">direction =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="hierarchical-deconvolution-using-scrnaseq-defined-macrophage-subtypes" class="level2">
<h2 class="anchored" data-anchor-id="hierarchical-deconvolution-using-scrnaseq-defined-macrophage-subtypes">7 Hierarchical Deconvolution using scRNAseq-defined macrophage subtypes</h2>
<section id="extracting-macrophage-specific-expression-from-spatial-domains" class="level3">
<h3 class="anchored" data-anchor-id="extracting-macrophage-specific-expression-from-spatial-domains">7.1 Extracting Macrophage-specific expression from spatial domains</h3>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>spatial_macs <span class="ot">&lt;-</span> <span class="fu">subset</span>(merged, <span class="at">subset =</span> main_annotation <span class="sc">==</span> <span class="st">'Macrophages'</span> <span class="sc">&amp;</span> tissue <span class="sc">==</span> <span class="st">"PDAC"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>spatial_macs_aggregate <span class="ot">&lt;-</span> <span class="fu">AggregateExpression</span>(spatial_macs, <span class="at">assays =</span> <span class="st">'Spatial'</span>, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"sample_id"</span>), <span class="at">return.seurat =</span> T)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>ct_mtx <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">t</span>(<span class="fu">GetAssayData</span>(spatial_macs_aggregate, <span class="at">assay =</span> <span class="st">'Spatial'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>intersect_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(ct_mtx), <span class="fu">colnames</span>(bp_scRef<span class="sc">$</span>sc.dat))</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>myPrism <span class="ot">&lt;-</span> <span class="fu">new.prism</span>(</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference=</span>bp_scRef<span class="sc">$</span>sc.dat[,intersect_genes],</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">mixture=</span>ct_mtx[,intersect_genes],</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.type.labels =</span> bp_scRef<span class="sc">$</span>metadata<span class="sc">$</span>main_annotation_scvi,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.state.labels =</span> bp_scRef<span class="sc">$</span>metadata<span class="sc">$</span>main_annotation_scvi_tissue,</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">key =</span> <span class="st">'Tumor_Epithelial'</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>bp_res <span class="ot">&lt;-</span> <span class="fu">run.prism</span>(myPrism, <span class="at">n.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>spatial_macs_aggregate[[<span class="st">'macexp'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.exp</span>(bp_res, <span class="at">state.or.type =</span> <span class="st">'type'</span>, <span class="at">cell.name =</span> <span class="st">'Macrophages'</span>))))</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>spatial_macs_aggregate[[<span class="st">'celltype_fractions'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.fraction</span>(bp_res, <span class="at">which.theta =</span> <span class="st">'final'</span>, <span class="at">state.or.type =</span> <span class="st">'type'</span>)), <span class="at">digits =</span> <span class="dv">2</span>))</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>spatial_macs_aggregate<span class="sc">@</span>assays<span class="sc">$</span>macexp<span class="sc">$</span>data <span class="ot">&lt;-</span> <span class="fu">vst</span>(<span class="fu">as.matrix</span>(<span class="fu">GetAssayData</span>(spatial_macs_aggregate, <span class="at">assay =</span> <span class="st">'macexp'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes-1" class="level3">
<h3 class="anchored" data-anchor-id="hierarchical-deconvolution-using-scrnaseq-defined-fibroblast-subtypes-1">7.2 Hierarchical Deconvolution using scRNAseq-defined fibroblast subtypes</h3>
<section id="generating-macrophages-scrnaseq-bayesprism-reference" class="level4">
<h4 class="anchored" data-anchor-id="generating-macrophages-scrnaseq-bayesprism-reference">7.2.1 Generating macrophages scRNAseq BayesPrism Reference</h4>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>macs <span class="ot">&lt;-</span> <span class="fu">subset</span>(macs, <span class="at">subset =</span> tissue <span class="sc">!=</span> <span class="st">'AdjNormal'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> macs<span class="sc">@</span>meta.data</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>macs_subtypes_tissue <span class="ot">&lt;-</span> <span class="fu">paste0</span>(metadata<span class="sc">$</span>macs_subtypes, <span class="st">"_"</span>, metadata<span class="sc">$</span>tissue)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>sc.dat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">GetAssayData</span>(macs, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered <span class="ot">&lt;-</span> <span class="fu">cleanup.genes</span>(<span class="at">input=</span><span class="fu">as.matrix</span>(sc.dat),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">species=</span><span class="st">"hs"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">gene.group=</span><span class="fu">c</span>( <span class="st">"Rb"</span>,<span class="st">"Mrp"</span>,<span class="st">"other_Rb"</span>,<span class="st">"chrM"</span>,<span class="st">"MALAT1"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span>) ,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">exp.cells=</span><span class="dv">5</span>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>sc.dat.filtered.pc <span class="ot">&lt;-</span>  <span class="fu">select.gene.type</span>(sc.dat.filtered, <span class="at">gene.type =</span> <span class="st">"protein_coding"</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>macs_scRef <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">sc.dat =</span> sc.dat.filtered.pc, <span class="at">metadata =</span> metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="running-bayesprism-1" class="level4">
<h4 class="anchored" data-anchor-id="running-bayesprism-1">7.2.2 Running BayesPrism</h4>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ct_mtx <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">t</span>(<span class="fu">GetAssayData</span>(spatial_macs_aggregate, <span class="at">assay =</span> <span class="st">'macexp'</span>, <span class="at">layer =</span> <span class="st">'counts'</span>)))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>intersect_genes <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(ct_mtx), <span class="fu">colnames</span>(macs_scRef<span class="sc">$</span>sc.dat))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>myPrism <span class="ot">&lt;-</span> <span class="fu">new.prism</span>(</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference=</span>macs_scRef<span class="sc">$</span>sc.dat[,intersect_genes],</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">mixture=</span>ct_mtx[,intersect_genes],</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">input.type=</span><span class="st">"count.matrix"</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.type.labels =</span> macs_scRef<span class="sc">$</span>metadata<span class="sc">$</span>macs_subtypes,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.state.labels =</span> macs_scRef<span class="sc">$</span>metadata<span class="sc">$</span>macs_subtypes_tissue,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">key =</span> <span class="cn">NULL</span>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>macs_bp_res <span class="ot">&lt;-</span> <span class="fu">run.prism</span>(myPrism, <span class="at">n.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="visualizing-fibroblast-subtype-fractions-1" class="level4">
<h4 class="anchored" data-anchor-id="visualizing-fibroblast-subtype-fractions-1">7.2.3 Visualizing Fibroblast Subtype Fractions</h4>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>spatial_macs_aggregate[[<span class="st">'subtype_fractions'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="fu">round</span>(<span class="fu">t</span>(<span class="fu">get.fraction</span>(macs_bp_res, <span class="at">which.theta =</span> <span class="st">'final'</span>, <span class="at">state.or.type =</span> <span class="st">'type'</span>)), <span class="at">digits =</span> <span class="dv">2</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(spatial_macs_aggregate) <span class="ot">&lt;-</span>  <span class="st">'subtype_fractions'</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spatial_macs_aggregate, <span class="st">"outputs/Macs_Analysis/macs_subtypes_deconvolution.qs"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>plot_df <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(spatial_macs_aggregate, <span class="at">assay =</span> <span class="st">'subtype_fractions'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"mac_subtype"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>mac_subtype, <span class="at">values_to =</span> <span class="st">'fraction'</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder_within</span>(<span class="at">x =</span> mac_subtype, <span class="at">within =</span> mac_subtype, <span class="at">by =</span> fraction), <span class="at">y =</span> fraction)) <span class="sc">+</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_reordered</span>() <span class="sc">+</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Macrophage Subtype"</span>, <span class="at">y =</span> <span class="st">"Fraction"</span>, <span class="at">title =</span> <span class="st">"Macrophage Subtype Fractions in Spatial Domains"</span>) <span class="sc">+</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>



</section>
</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-bayesprism" class="csl-entry" role="listitem">
Chu, Tinyi, Zhong Wang, Dana Peâ€™er, and Charles G Danko. 2022. <span>â€œCell Type and Gene Expression Deconvolution with BayesPrism Enables Bayesian Integrative Analysis Across Bulk and Single-Cell RNA Sequencing in Oncology.â€</span> <em>Nature Cancer</em> 3 (4): 505â€“17.
</div>
<div id="ref-DESeq2" class="csl-entry" role="listitem">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>â€œModerated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.â€</span> <em>Genome Biology</em> 15 (12): 550.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>